### DISCUSSION TITLE
About the LoRA merging (Video models)

### DISCUSSION THREAD

**Discussion by dorpxam** [kohya-ss/musubi-tuner#316] [Q&A]
I want to experiment the merging of two or more Hunyuan Lora. The idea is, for a character LoRA, to split the training in different shorts concepts (walking, dancing, some moves ...) datasets limited in size.

I don't know if it's a good idea and if can do better result ?

Is a script exists to merge LoRAs together, not merging to the transformer model ?

Thanks a lot for your advice.

**Comment by dorpxam** [comment]
No one ? Maybe it's a stupid question I know. I'm newbie in this domain and my knowledge is very limited. 

I will attempt to reuse pre-trained LoRa using `--network_weights` but I'm afraid because that mean that the weights will be greatly modified by the optimizer while a merge of two good trained LoRa by mean of the weight seem a better solution.

That said, maybe it's a bad idea to split the training in multiple stage. I hope someone had the experience to guide me in the right direction. I ask for nothing but good advice.

**Reply by Sarania** [reply]
I don't think it's a stupid question but I don't have a good answer for you. It should be possible to merge LoRA the way you want (though whether it will give you the results you are after is another question) but there are no easy scripts that I know of to do so. That said I can tell you that resuming with --network_weights prolly won't work, I trained a dataset with (not exact numbers but close) 16 examples of one subject and 8 of an adjacent subject for 2000 steps, it picked up the subject with 16 examples but not quite the one with 8, so I gave it some more steps just on those 8... but nah. I think having a balanced dataset from the start is prolly the way.

**Reply by dorpxam** [reply]
Thank's for your answer @Sarania. Yes, for the moment I experiment the `--network_weights` but I can't say if the result is good or not. The test I've done is about a pre-trained weights trained over a set of image, used as base for a post-training over videos. I can't really say that the current result (stopped @ 21 epochs for a avg.loss of 0.0518 for now) is good because of the training after the pre-training or not. 

But I can say that some LoRA on civitai are merge of others LoRAs. The creator say that he use a personal script to do the job and share a link to ... a paid space with really annoying ID verification. Boring!

If I knew which tensors in LoRA to mix, I would try to code a script myself. But I must admit that I don't have the knowledge level for that. I imagine that it should not be particularly complex, especially if the goal is to mix LoRAs based on the same character or concept, with the same training parameters, the same trigger words...

**Reply by Sarania** [reply]
> The creator say that he use a personal script to do the job and share a link to ... a paid space with really annoying ID verification. Boring!

That really sucks and yeah I've seen mention of the same on Civit but a lot of people don't like to share. 🙁 That's actually part of why https://github.com/Sarania/blissful-tuner/ exists - it really is just my own personal scripts and extensions to Musubi cleaned up and made nice for sharing. I don't have anything for merging LoRA yet specifically, but I do at least have a model utility that can show you the keys and shapes inside most types of transformer models including LoRA. Example:

```
python blissful_tuner/model_utility.py --input /home/blyss/projects/art/BlissfulModels/WanKisses/WanKisses.safetensors --inspect
```
That will show you the key and tensor structure inside of a given LoRA which might help you decide what to merge. Honestly I've been considering making a LoRA utility script for Blissful Tuner anyway with the capability to merge, derank, etc. I can't promise a timeframe though because when I do it just stresses me out lol.

That said if you wanna have a go at it yourself generally there are two LoRA keys for each model key. The actual key naming can vary but for instance Musubi Tuner LoRA will have key names like
```
lora_unet_blocks_9_cross_attn_q.lora_down.weight
lora_unet_blocks_9_cross_attn_q.lora_up.weight
```
vs diffusion-pipe:
```
diffusion_model.blocks.9.cross_attn.q.lora_A.weight
diffusion_model.blocks.9.cross_attn.q.lora_B.weight
```
so you'll need to account for that but generally you are gonna wanna merge them 1 to 1. Whether the best way is to average the tensors together or do some other maths I can't say I would have to research that. You'll also wanna make sure they are the same dtype - potentially just upcast the merge to float32 and save the results as whatever you dtype you like at the end.

**Reply by dorpxam** [reply]
Oh! Very interesting!

1) Yes. Making money with open-source project is a shame. More of that, sharing good stuff allows better propagation of derived material. I will never understand guys who keep this kind of stuff like treasure.

2) The LoRA inspector : very cool and useful.

3) If you want to work on scripts, that can be cool. Anyway, I'm not the kind of guy who makes others work and waits for everything to fall into his lap. So, I will share my experience and the progress of my experiments here. If you can do the same, that's cool.

4) Yes, the idea is to have the possibility of merging like [mergekit](https://github.com/arcee-ai/mergekit) for the LLM. The real problem is to know exactly what weights to merge to not corrupt the final LoRA. In the same spirit, the choice of the blocks to affect can be crucial to not degrade the LoRA. Single blocks only, double blocks only, both ? [here](https://github.com/facok/ComfyUI-HunyuanVideoMultiLora/tree/main) the nodes allow to affect only the double blocks. 

Maybe the idea will be to have a script that allow to merge N LoRAs, with different strength, different merging method (mean, geometric mean, ... maybe some more complex method like SVD) and the possibility to affect all or parts of the blocks.

If you have information about the blocks inside the Hunyuan LoRA, that can be cool. That remind me the FLUX training on few layers only proposed by TheLastBen [here](https://www.reddit.com/r/StableDiffusion/comments/1f523bd/good_flux_loras_can_be_less_than_45mb_128_dim/)
